### PR TITLE
Feature/contrast enhancement algorithms

### PR DESCRIPTION
**Pull request by danielbabiak**

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -13,7 +13,7 @@ environment:
 configuration:
   - Debug
   - Release
-version: 3.4.0.{build}
+version: 3.5.0.{build}
```
Comment (gmrukwa) [inline appveyor.yml]: Depending on which will be merged first, you may need to update it once more - @Sand3r- is setting the same version in his review #11.

File: src/Spectre.libHeatmapDataScaling/HeatmapDataScalingAlgorithm.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+limitations under the License.
+*/
+
+#pragma once
+#include <span.h>
+#include <vector>
+
+namespace Spectre::libHeatmapDataScaling
+{
+    /// <summary>
```
Comment (gmrukwa) [inline src/Spectre.libHeatmapDataScaling/HeatmapDataScalingAlgorithm.h]: Tabs / spaces are mixed in this file. We are generally using spaces.
Content of namespace should not be extra-indented.

File: src/Spectre.libHeatmapDataScaling/HeatmapDataScalingAlgorithm.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+#include <span.h>
+#include <vector>
+
+namespace Spectre::libHeatmapDataScaling
```
Comment (gmrukwa) [inline src/Spectre.libHeatmapDataScaling/HeatmapDataScalingAlgorithm.h]: Please, normalize namespace to current structure. Maybe `spectre::visualization`?

File: src/Spectre.libHeatmapDataScaling/Normalization.cpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "Normalization.h"
+
+namespace Spectre::libHeatmapDataScaling
+{
+	Normalization::Normalization(const int _min, const int _max) : min(_min), max(_max) { };
```
Comment (gmrukwa) [inline src/Spectre.libHeatmapDataScaling/Normalization.cpp]: Inconsistent naming. Function arguments should be `lowerCamelCase`.

File: src/Spectre.libHeatmapDataScaling/Normalization.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+	std::vector<double> Normalization::scaleData(const gsl::span<double> intensities)
+	{
+        std::vector<double> newIntensities;
+		newIntensities.reserve(intensities.size());
```
Comment (gmrukwa) [inline src/Spectre.libHeatmapDataScaling/Normalization.cpp]: Indents are broken.

File: src/Spectre.libHeatmapDataScaling/Normalization.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		double const oldMin = *min_element(std::begin(intensities), std::end(intensities));
+		double const oldMax = *max_element(std::begin(intensities), std::end(intensities));
+
+		for (int i = 0; i< intensities.size(); i++)
+		{
+			newIntensities.push_back(((intensities[i] - oldMin) * (max - min) / (oldMax - oldMin)) + min);
```
Comment (gmrukwa) [inline src/Spectre.libHeatmapDataScaling/Normalization.cpp]: We may simply initialize vector with proper length and then write like any other array: `newIntensities[i] = ...`.

Computations like `max - min`, `oldMax - oldMin` are repeated each iteration. They could be extracted before the loop and named properly. Or even extracted as a kind of a normalization constant.

File: src/Spectre.libHeatmapDataScaling/Normalization.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+#include <vector>
+#include "HeatmapDataScalingAlgorithm.h"
```
Comment (gmrukwa) [inline src/Spectre.libHeatmapDataScaling/Normalization.h]: We need directory prefix in such case, when we include header from header (NuGet packaging purposes).

File: src/Spectre.libHeatmapDataScaling/Normalization.h [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    /// </summary>
+	class Normalization : public HeatmapDataScalingAlgorithm
+	{
+	public:
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Normalization"/> class.
+        /// </summary>
+        /// <param name="_min">Minimum value of the range.</param>
+        /// <param name="_max">Maximum value of the range.</param>
+		Normalization(const int _min = 0, const int _max = 100);
```
Comment (gmrukwa) [inline src/Spectre.libHeatmapDataScaling/Normalization.h]: Inconsistent variable naming.

File: src/Spectre.libHeatmapDataScaling/Normalization.h [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+        /// <summary>
+        /// Method for data normalization to the given range.
+        /// </summary>
+        /// <param name="intensities">Vector of floating point values representing intensities for each point on the heatmap.</param>
+        /// <returns>Vector of normalized floating point values representing intensities for each point on the heatmap.</returns>
+		std::vector<double> scaleData(const gsl::span<double> intensities) override;
+	private:
+		const int min, max;
```
Comment (gmrukwa) [inline src/Spectre.libHeatmapDataScaling/Normalization.h]: `min` and `max` are missing any information about the context. I know this is simple class & functionality, but we all have sometimes bad day, when it's hard to grasp the context on the first look. I think these could be renamed to be more expressive.

File: src/Spectre.libHeatmapDataScaling/Spectre.libHeatmapDataScaling.vcxproj [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'" Label="Configuration">
+    <ConfigurationType>StaticLibrary</ConfigurationType>
+    <UseDebugLibraries>false</UseDebugLibraries>
+    <PlatformToolset>v140</PlatformToolset>
+    <WholeProgramOptimization>true</WholeProgramOptimization>
+    <CharacterSet>MultiByte</CharacterSet>
+  </PropertyGroup>
+  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.props" />
+  <ImportGroup Label="ExtensionSettings">
```
Comment (gmrukwa) [inline src/Spectre.libHeatmapDataScaling/Spectre.libHeatmapDataScaling.vcxproj]: Why do we need this and next empty property group?

File: appveyor.yml (hunk 1 above)
Comment (Sand3r-) [reply appveyor.yml]: You can bump it up to 6.

**Comment by gmrukwa** [comment]
[Prepare tests for heatmap scaling algorithms and refactoring](https://trello.com/c/RamXoMUZ/353-prepare-tests-for-heatmap-scaling-algorithms-and-refactoring)

File: src/Spectre.libHeatmapDataScaling/HeatmapDataScalingAlgorithm.h [hunk 11]
```diff
@@ -18,24 +18,24 @@ limitations under the License.
 #include <span.h>
 #include <vector>
 
-namespace Spectre::libHeatmapDataScaling
+namespace Spectre::visualization
```
Comment (gmrukwa) [inline src/Spectre.libHeatmapDataScaling/HeatmapDataScalingAlgorithm.h]: Root namespace is lowercase, like `spectre`.

File: src/Spectre.libHeatmapDataScaling/Normalization.cpp [hunk 12]
```diff
@@ -14,23 +14,24 @@ See the License for the specific language governing permissions and
...
-        std::vector<double> newIntensities;
-		newIntensities.reserve(intensities.size());
-		double const oldMin = *min_element(std::begin(intensities), std::end(intensities));
-		double const oldMax = *max_element(std::begin(intensities), std::end(intensities));
+std::vector<double> Normalization::scaleData(const gsl::span<double> intensities)
+{
+    std::vector<double> newIntensities(intensities.size());
+    double const oldMin = *min_element(std::begin(intensities), std::end(intensities));
+    double const oldMax = *max_element(std::begin(intensities), std::end(intensities));
+    double const differenceBetweenOldMaxAndOldMin = oldMax - oldMin;
```
Comment (gmrukwa) [inline src/Spectre.libHeatmapDataScaling/Normalization.cpp]: Isn't it just `oldRange`? :D

**Pull request merged**